CARRY: feat(RHOAIENG-31239): "Add networkpolicy support"

chipspeak opened the pull request:
## Why are these changes needed?
This PR adds `NetworkPolicy` support to ensure secure RayCluster pod isolation.

## Verification Steps
### Setup

* First, build this branch using the below command from within the `ray-operator` directory.

```shell
make docker-build IMG=network-policy-test:odh
```

* Now we can load that image into kind using:

```shell
kind load docker-image network-policy-test:odh --name network-policy-test
```

* We need to `ray-operator/config/manager/manager.yaml` to point to our new local image and to set the `imagePullPolicy` to never (this ensures that it falls back to the local image rather than pulling from a registry). You can do this by making the follow adjustment to line 34 and 35:

```shell
image: network-policy-test:odh
imagePullPolicy: Never
```

* We can then install the operator using the below command:

```shell
kubectl apply --server-side -k config/default
```

* Verify that the operator is in a running state by using the below command:


```shell
kubectl get pods | grep kuberay-operator
```

* We can now create a RayCluster using the sample included in this PR by running the below command (note the annotation applied in this example as it is required for the network policy):

```shell
kubectl apply -f ray-operator/raycluster-networkpolicy.yaml
```

* Once the RayCluster is up and running we can verify that a matching NetworkPolicy has been created by running the below command:

```shell
kubectl get networkpolicies -n default
```

* This should return `raycluster-kuberay-head` and `raycluster-kuberay-worker`. You can view the NetworkPolicy by running the below command:

```shell
kubectl describe networkpolicy <network_policy_name>
```

### Rule Verification

**Rule 1**

* A pod with a label matching the newly created RayCluster can access all ports. We can verify this by running the below command:

```shell
kubectl run test-rule1-intra --image=busybox --rm -i --restart=Never -n default --labels="ray.io/cluster=raycluster-kuberay" -- timeout 5 nc -zv raycluster-kuberay-head-svc 6379
```

* This should succeed as we’ve applied the required label to the pod, granting it access to secure ports. Lets further verify this by attempting to access another secure port via a pod with the same label:

```shell
kubectl run test-rule1-metrics --image=busybox --rm -i --restart=Never -n default --labels="ray.io/cluster=raycluster-kuberay" -- timeout 5 nc -zv raycluster-kuberay-head-svc 8080
```

* This should work once again. We can verify the expected failure when the label is missing by attempting to access GCS (port 6379\) again but without a label:

```shell
kubectl run test-rule1-intra --image=busybox --rm -i --restart=Never -n default -- timeout 5 nc -zv raycluster-kuberay-head-svc 6379
```

* This will have failed as we don’t have permission to access that port without the label.

**Rule 2** 

* We’re restricting communication with the Ray dashboard and client to pods within the same namespace. We can verify this by running the following to create a pod and query the dashboard.

```shell
kubectl run test-same-namespace --image=busybox --rm -i --restart=Never -- \
  timeout 1 wget -qO- raycluster-kuberay-head-svc:8265
```

* This should succeed and return the HTML response from the dashboard as seen below.

```html
<!doctype html><html lang="en"><head><meta charset="utf-8"/><link rel="shortcut icon" href="./favicon.ico"/><meta name="viewport" content="width=device-width,initial-scale=1"/><title>Ray Dashboard</title><script defer="defer" src="./static/js/main.04e1bfe3.js"></script><link href="./static/css/main.388a904b.css" rel="stylesheet"></head><body><noscript>You need to enable JavaScript to run this app.</noscript><div id="root"></div></body></html>
```

* We can further verify this by doing the same, but with a pod from a different namespace. Run the below commands in order to achieve this.

```shell
kubectl create namespace test-namespace

kubectl run test-diff-namespace --image=busybox --rm -i --restart=Never -n test-namespace -- timeout 2 wget -qO- raycluster-kuberay-head-svc.default:8265
```

* This time, we should timeout as this pod is not permitted to communicate with the Ray Head’s dashboard as it is from an external namespace.

**Rule 3**

* The KubeRay operator should also have appropriate permissions to communicate with the RayCluster pods. This requires that the operator is within an expected namespace (dynamically retrieved by the controller) and has the `app.kubernetes.io/name=kuberay` label. We can verify this by running the below command.

```shell
kubectl run test-rule3-operator --image=busybox --rm -i --restart=Never -n default --labels="app.kubernetes.io/name=kuberay" -- timeout 5 nc -zv raycluster-kuberay-head-svc 8265
```

* This will pass implicitly due to rule 2, regardless of the label as the kuberay operator is within the same namespace as the RayCluster we created. We’ll need to create a RayCluster in the test-namespace we created earlier by running:

```shell
kubectl apply -f ray-operator/raycluster-networkpolicy.yaml -n test-namespace
```

* Now we can create a pod with the KubeRay operator’s label in our default namespace and attempt to access this pod:

```shell
kubectl run test-cross-ns-operator --image=busybox --rm -i --restart=Never -n default --labels="app.kubernetes.io/name=kuberay" -- timeout 5 nc -zv raycluster-kuberay-head-svc.test-namespace 8265
```

* This should succeed due to rule 3\. Let’s try the same but minus the KubeRay operator label:

```shell
kubectl run test-cross-ns-no-label --image=busybox --rm -i --restart=Never -n default -- timeout 5 nc -zv raycluster-kuberay-head-svc.test-namespace 8265
```

* This should fail as the label is required for cross-namespace communication. This is effectively what rule 3 adds to the rule 2 logic.

**Rule 4**

* We permit access to the monitoring port (8080) from two common monitoring namespaces: openshift-monitoring and prometheus. To verify this access we first need to create these namespaces by running:

```shell
kubectl create namespace openshift-monitoring && kubectl create namespace prometheus
```

* Now we can test the openshift namespace’s access by running:

```shell
kubectl run test-rule4-openshift --image=busybox --rm -i --restart=Never -n openshift-monitoring -- timeout 5 nc -zv raycluster-kuberay-head-svc.default 8080
```

* This should succeed. We can also verify the prometheus namespace’s access by running:

```shell
kubectl run test-rule4-prometheus --image=busybox --rm -i --restart=Never -n prometheus -- timeout 5 nc -zv raycluster-kuberay-head-svc.default 8080
```

* This should once again succeed. We can also confirm that this access is limited to the above by attempting to access port 8080 from a different, non-monitoring namespace by running:

```shell
kubectl run test-rule4-blocked --image=busybox --rm -i --restart=Never -n test-namespace -- timeout 5 nc -zv raycluster-kuberay-head-svc.default 8080
```

* This should fail as this pod is from a non-monitoring namespace (and not covered by Rule 2 as it is not in the same namespace as the RayCluster). We can verify that the rule limits the monitoring namespaces access appropriately by running:

```shell
kubectl run test-rule4-wrong-port --image=busybox --rm -i --restart=Never -n openshift-monitoring -- timeout 5 nc -zv raycluster-kuberay-head-svc.default 6379
```

* This should fail as these monitoring namespaces are only permitted to access port 8080\.

**Rule 5**

* We permit access to port 8443 from any pod but Ray won’t be listening on this port unless mTLS has been configured for this particular RayCluster. We can verify this by running:

```shell
kubectl run test-rule5-non-mtls --image=busybox --rm -i --restart=Never -n test-namespace -- timeout 2 nc -zv raycluster-kuberay-head-svc.default 8443
```

* This should timeout as this RayCluster is not mTLS enabled and as a result, is not listening on port 8443\. We can further illustrate this by checking the head pods NetworkPolicy:

```shell
kubectl describe networkpolicy raycluster-kuberay
```

* In the output of the above, you should be able to observe that in the final rule, only port 8443 is exposed as seen below. 

```shell
    To Port: 8443/TCP
    From: <any> (traffic not restricted by source)
```

* We can demonstrate how a RayCluster with mTLS enabled is in fact listening on this port by applying a CR with mTLS enabled. In this instance, we don’t have an mTLS controller to automatically add a container port of 8443 so we’ll do it manually by running:

```shell
kubectl apply -f - <<EOF
apiVersion: ray.io/v1
kind: RayCluster
metadata:
  name: raycluster-kuberay-mtls
  annotations:
    odh.ray.io/secure-trusted-network: "true"
spec:
  rayVersion: '2.46.0'
  headGroupSpec:
    rayStartParams: {}
    template:
      spec:
        containers:
        - name: ray-head
          image: rayproject/ray:2.46.0
          env:
          - name: RAY_USE_TLS
            value: "1"
          - name: RAY_TLS_SERVER_CERT
            value: "/etc/tls/server.crt"
          - name: RAY_TLS_SERVER_KEY
            value: "/etc/tls/server.key"
          resources:
            limits:
              cpu: 1
              memory: 2G
            requests:
              cpu: 1
              memory: 2G
          ports:
          - containerPort: 6379
            name: gcs-server
          - containerPort: 8265
            name: dashboard
          - containerPort: 10001
            name: client
          - containerPort: 8443
            name: mtls
  workerGroupSpecs:
  - replicas: 1
    minReplicas: 1
    maxReplicas: 5
    groupName: workergroup
    rayStartParams: {}
    template:
      spec:
        containers:
        - name: ray-worker
          image: rayproject/ray:2.46.0
          env:
          - name: RAY_USE_TLS
            value: "1"
          - name: RAY_TLS_SERVER_CERT
            value: "/etc/tls/server.crt"
          - name: RAY_TLS_SERVER_KEY
            value: "/etc/tls/server.key"
          resources:
            limits:
              cpu: 1
              memory: 1G
            requests:
              cpu: 1
              memory: 1G
EOF
```

* NOTE: this is only verifiable from the NetworkPolicy and port exposure perspective. This hinges on mTLS being configured as part of KubeRay in the future. While this will succeed in creating an appropriate network policy, the pods will fail as there will be no certificates available.  
* We can verify that the ports are exposed correctly by running:

```shell
kubectl describe networkpolicy raycluster-kuberay-mtls-head
```

* In the output, you should be able to observe the final ruleset and see that port 10001 is now also exposed as seen below. This is expected when mTLS is configured.

```shell
    To Port: 8443/TCP
    To Port: 10001/TCP
    From: <any> (traffic not restricted by source)
```


## Related issue number

[Jira Link](https://issues.redhat.com/browse/RHOAIENG-31239)

_NOTE_: While this is marked for upstream, it is being added in ODH to ensure successful CFO removal for RHOAI 3.x.

## Checks

- [ ] I've made sure the tests are passing.
- Testing Strategy
  - [ ] Unit tests
  - [ ] Manual tests
  - [ ] This PR is not tested :(


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added NetworkPolicy management for RayCluster resources with annotation-based activation (secure-trusted-network)
  * Added network policy event tracking for created, deleted, and failed operations

* **Configuration**
  * Updated RBAC permissions to include NetworkPolicy resources under the networking.k8s.io API group
  * Added POD_NAMESPACE environment variable to operator deployment

* **Tests**
  * Added comprehensive integration and unit test coverage for NetworkPolicy controller
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

chipspeak committed:
CARRY: feat(RHOAIENG-31239): "Add networkpolicy support"

Signed-off-by: Pat O'Connor <paoconno@redhat.com>

chipspeak commented:
/hold

kryanbeane commented on ray-operator/config/manager/manager.yaml:
```diff
@@ -80,4 +80,9 @@ spec:
           # environment variable is not set, requeue after the default value (300).
           # - name: RAYCLUSTER_DEFAULT_REQUEUE_SECONDS_ENV
           #   value: "300"
+          # Required for NetworkPolicy feature when operator is NOT deployed in 'ray-system' namespace
```
Does this need to be uncommented by default then if this is in ODH or is it ok commented out

chipspeak replied:
Yeah you'd be spot on, this is legacy from the upstream. I'll sort that now. I guess we'll have to document this and the various other places where we'll need to be mindful of during syncs!

kryanbeane reviewed (commented):
Only one question from me below. Otherwise it looks great. I will run verification steps now!

kryanbeane reviewed (approved):
Verified this and all worked as expected, looks great Pat well done!

laurafitzgerald commented on helm-chart/kuberay-operator/README.md:
```diff
@@ -165,6 +165,8 @@ spec:
 | featureGates[1].name | string | `"RayJobDeletionPolicy"` |  |
 | featureGates[1].enabled | bool | `false` |  |
+| featureGates[2].name | string | `"RayClusterNetworkPolicy"` |  |
+| featureGates[2].enabled | bool | `false` |  |
```
@chipspeak can we turn this on by default here given it will be the default in odh.

laurafitzgerald commented on helm-chart/kuberay-operator/values.yaml:
```diff
@@ -88,6 +88,8 @@ featureGates:
 - name: RayJobDeletionPolicy
   enabled: false
+- name: RayClusterNetworkPolicy
+  enabled: false
```
same here

laurafitzgerald commented on ray-operator/pkg/features/features.go:
```diff
@@ -33,6 +36,7 @@ func init() {
 var defaultFeatureGates = map[featuregate.Feature]featuregate.FeatureSpec{
 	RayClusterStatusConditions: {Default: true, PreRelease: featuregate.Beta},
 	RayJobDeletionPolicy:       {Default: false, PreRelease: featuregate.Alpha},
+	RayClusterNetworkPolicy:    {Default: false, PreRelease: featuregate.Alpha},
```
same here

chipspeak commented:
> Verified this and all worked as expected, looks great Pat well done!

Thanks so much, Bryan! I'll have to account for prometheus too based on last weeks conversation so I'll revisit this and request one more glance at some point if that's ok!

kryanbeane reviewed (approved):
re-verified and reviewed, lgtm

chipspeak commented:
I've updated the PR to move to an annotation based approach and updated the verification steps accordingly if folks can re-review cc @laurafitzgerald @CathalOConnorRH @pawelpaszki @LilyLinh @kryanbeane

chipspeak commented:
@LilyLinh @pawelpaszki Think I got to the bottom of it. We dynamically pulled the KubeRay operator's namespace but it was falling back to `ray-system` because the `POD_NAMESPACE` var was never being set. I've added it to the `manager.yaml`. I've validated that this resolves the issue. Ignore the web-hook stuff for now. Was just easier to work on both tickets simultaneously.

pawelpaszki commented on ray-operator/config/manager/manager.yaml:
```diff
@@ -27,10 +27,10 @@ spec:
 #        - --enable-leader-election
-        image: kuberay/operator
-        imagePullPolicy: IfNotPresent
+        image: network-policy-test:odh
```
this and next line should only be for testing purposes, right (e.g. not committed)?

chipspeak replied:
committed to make verification easier. these aren't pushed anywhere, it's just the name I've used in the verification steps so it would make the reviewers life easier. They'll be reverted before merge etc

pawelpaszki reviewed (commented):
I can confirm that the verification steps work as expected

laurafitzgerald commented on ray-operator/controllers/ray/networkpolicy_controller.go:
```diff
@@ -0,0 +1,460 @@
+
+// getKubeRayNamespaces returns the list of KubeRay operator namespaces
+func (r *NetworkPolicyController) getKubeRayNamespaces(_ context.Context) []string {
+	operatorNamespace := os.Getenv("POD_NAMESPACE")
```
is it possible that this entry can be changed at all?

laurafitzgerald commented on ray-operator/controllers/ray/networkpolicy_controller.go:
```diff
@@ -0,0 +1,368 @@
+func (r *NetworkPolicyController) createOrUpdateNetworkPolicy(ctx context.Context, instance *rayv1.RayCluster, networkPolicy *networkingv1.NetworkPolicy) error {
+	logger := ctrl.LoggerFrom(ctx).WithName("networkpolicy-controller")
+
+	// Set owner reference for garbage collection
+	if err := controllerutil.SetControllerReference(instance, networkPolicy, r.Scheme); err != nil {
+		return err
+	}
+
+	// Try to create the NetworkPolicy
+	if err := r.Create(ctx, networkPolicy); err != nil {
+		if errors.IsAlreadyExists(err) {
+			// NetworkPolicy exists, update it
+			existing := &networkingv1.NetworkPolicy{}
+			if err := r.Get(ctx, client.ObjectKeyFromObject(networkPolicy), existing); err != nil {
+				return err
+			}
+
+			// Update the existing NetworkPolicy
+			existing.Spec = networkPolicy.Spec
+			existing.Labels = networkPolicy.Labels
+
+			if err := r.Update(ctx, existing); err != nil {
+				r.Recorder.Eventf(instance, corev1.EventTypeWarning, string(utils.FailedToCreateNetworkPolicy),
+					"Failed to update NetworkPolicy %s/%s: %v", networkPolicy.Namespace, networkPolicy.Name, err)
+				return err
+			}
```
am i right in saying that we want full control over the network policies, including labels such that additional ones can't be added such that this could be bypassed?

laurafitzgerald commented on ray-operator/controllers/ray/networkpolicy_controller.go:
```diff
@@ -0,0 +1,460 @@
+			logger.Info("Successfully updated NetworkPolicy", "name", networkPolicy.Name)
+			r.Recorder.Eventf(instance, corev1.EventTypeNormal, string(utils.CreatedNetworkPolicy),
+				"Updated NetworkPolicy %s/%s", networkPolicy.Namespace, networkPolicy.Name)
+		} else {
```
this else could in theory be removed

laurafitzgerald commented on ray-operator/controllers/ray/networkpolicy_controller.go:
```diff
@@ -0,0 +1,460 @@
+				"Failed to create NetworkPolicy %s/%s: %v", networkPolicy.Namespace, networkPolicy.Name, err)
+			return err
+		}
+	} else {
```
same with this one.

laurafitzgerald commented on ray-operator/controllers/ray/networkpolicy_controller.go:
```diff
@@ -0,0 +1,460 @@
+	}
+
+	// Check if mTLS is enabled by looking for TLS configuration in RayCluster
+	if r.isMTLSEnabled(instance) {
```
This needs to be removed for now until we add the changes to the spec where the user has options around selecting the various configuarations. For now it'll be covered by the annotation

laurafitzgerald commented on ray-operator/controllers/ray/networkpolicy_controller.go:
```diff
@@ -0,0 +1,460 @@
+							{
+								Key:      corev1.LabelMetadataName,
+								Operator: metav1.LabelSelectorOpIn,
+								Values:   []string{"openshift-monitoring", "prometheus", "redhat-ods-monitoring"},
```
```suggestion
								Values:   []string{"openshift-monitoring", "redhat-ods-monitoring"},
```

laurafitzgerald commented on ray-operator/controllers/ray/networkpolicy_controller.go:
```diff
@@ -0,0 +1,460 @@
+// This looks for TLS-related environment variables or configuration
+func (r *NetworkPolicyController) isMTLSEnabled(instance *rayv1.RayCluster) bool {
+	// Check head group for TLS environment variables
+	if r.checkContainersForMTLS(instance.Spec.HeadGroupSpec.Template.Spec.Containers) {
```
these checks wouldn't be valid anymore given that we configure it directly on the resources

laurafitzgerald commented on ray-operator/controllers/ray/networkpolicy_controller.go:
```diff
@@ -0,0 +1,460 @@
+}
+
+// isNetworkPolicyEnabled checks if NetworkPolicy is enabled for this RayCluster via annotation
+func (r *NetworkPolicyController) isNetworkPolicyEnabled(instance *rayv1.RayCluster) bool {
```
can we rename this function to be more generaic given the annotation is generic.

laurafitzgerald commented on ray-operator/controllers/ray/suite_test.go:
```diff
@@ -73,6 +74,9 @@ func TestAPIs(t *testing.T) {
 var _ = BeforeSuite(func(ctx SpecContext) {
 	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
 
+	// Note: NetworkPolicy controller now uses annotation-based activation instead of feature flag
```
can be removed

laurafitzgerald commented on ray-operator/raycluster-networkpolicy.yaml:
```diff
@@ -0,0 +1,54 @@
+apiVersion: ray.io/v1
```
this example could be made more generic.

laurafitzgerald reviewed (changes requested):
@chipspeak looks good already. A few comments/questions/requests inline

LilyLinh commented:
I did all verification steps and it matched as described.

laurafitzgerald reviewed (approved):
Two follow up changes will be made in https://github.com/opendatahub-io/kuberay/pull/128
